files/archive.py

```diff
@@ -0,0 +1,385 @@
+    description: Any files that were missing from the source.
+    type: list
+    returned: success
+archived:
+    description: Any files that were compressed or added to the archive.
+    type: list
+    returned: success
+arcroot:
+    description: The archive root.
+    type: string
+expanded_paths:
+    description: The list of matching paths from paths argument.
+    type: list
+'''
+
+import stat
+import os
+import errno
+import glob
+import shutil
+import gzip
+import bz2
+import filecmp
+import zipfile
+import tarfile
+
+def main():
+    module = AnsibleModule(
+        argument_spec = dict(
+            path = dict(type='list', required=True),
+            compression = dict(choices=['gz', 'bz2', 'zip'], default='gz', required=False),
+            dest = dict(required=False),
+            remove = dict(required=False, default=False, type='bool'),
+        ),
+        add_file_common_args=True,
+        supports_check_mode=True,
+    )
+
+    params = module.params
+    paths = params['path']
+    dest = params['dest']
+    remove = params['remove']
+    expanded_paths = []
+    compression = params['compression']
+    globby = False
+    changed = False
+    state = 'absent'
+
+    # Simple or archive file compression (inapplicable with 'zip' since it's always an archive)
+    archive = False
+    successes = []
+
+    for i, path in enumerate(paths):
+        path = os.path.expanduser(path)
+
+        # Detect glob-like characters
+        if any((c in set('*?')) for c in path):
+            expanded_paths = expanded_paths + glob.glob(path)
+        else:
+            expanded_paths.append(path)
+
+    if len(expanded_paths) == 0:
+        return module.fail_json(path=', '.join(paths), expanded_paths=', '.join(expanded_paths), msg='Error, no source paths were found')
+
+    # If we actually matched multiple files or TRIED to, then
+    # treat this as a multi-file archive
+    archive = globby or os.path.isdir(expanded_paths[0]) or len(expanded_paths) > 1
+
+    # Default created file name (for single-file archives) to
+    # <file>.<compression>
+    if dest:
+        dest = os.path.expanduser(dest)
+    elif not archive:
+        dest = '%s.%s' % (expanded_paths[0], compression)
+
+    # Force archives to specify 'dest'
+    if archive and not dest:
+        module.fail_json(dest=dest, path=', '.join(paths), msg='Error, must specify "dest" when archiving multiple files or trees')
+
+    archive_paths = []
+    missing = []
+    exclude = []
+    arcroot = ''
+
+    for path in expanded_paths:
+        # Use the longest common directory name among all the files
+        # as the archive root path
+        if arcroot == '':
+            arcroot = os.path.dirname(path) + os.sep
+        else:
+            for i in xrange(len(arcroot)):
+                if path[i] != arcroot[i]:
+                    break
+
+            if i < len(arcroot):
+                arcroot = os.path.dirname(arcroot[0:i+1])
+
+            arcroot += os.sep
```

Comment:
xrange is not python3 compatible.  Not sure this does the right thing if, for example, you have ```expanded_paths = ['/srv/one-two/', '/srv/one-three']```

I think it will end up with ```arcroot = '/one-t/'```.  Probably need to write some code based around recursively calling ```os.path.split()``` or (a little hackier but usually correct) ```arcroot.split(os.path.sep)```